enterprise/backend/test/metabase_enterprise/advanced_permissions/driver/impersonation_test.clj

```diff
@@ -0,0 +1,94 @@
+(ns metabase-enterprise.advanced-permissions.driver.impersonation-test
+  (:require
+   [clojure.java.jdbc :as jdbc]
+   [clojure.test :refer :all]
+   [metabase-enterprise.advanced-permissions.api.util-test
+    :as advanced-perms.api.tu]
+   [metabase-enterprise.advanced-permissions.driver.impersonation
+    :as impersonation]
+   [metabase.driver.postgres-test :as postgres-test]
+   [metabase.driver.sql-jdbc.connection :as sql-jdbc.conn]
+   [metabase.models.database :refer [Database]]
+   [metabase.public-settings.premium-features-test
+    :as premium-features-test]
+   [metabase.server.middleware.session :as mw.session]
+   [metabase.sync :as sync]
+   [metabase.test :as mt]
+   [toucan2.tools.with-temp :as t2.with-temp]))
+
+(deftest connection-impersonation-role-test
+  (testing "Returns nil when no impersonations are in effect"
+    (is (nil? (@#'impersonation/connection-impersonation-role (mt/db)))))
+
+  (testing "Correctly fetches the impersonation when one is in effect"
+    (advanced-perms.api.tu/with-impersonations {:impersonations [{:db-id (mt/id) :attribute "impersonation_attr"}]
+                                                :attributes     {"impersonation_attr" "impersonation_role"}}
+      (is (= "impersonation_role"
+             (@#'impersonation/connection-impersonation-role (mt/db))))))
+
+  (testing "Throws exception if multiple conflicting impersonations are in effect"
+    ;; Use nested `with-impersonations` macros so that different groups are used
+    (advanced-perms.api.tu/with-impersonations {:impersonations [{:db-id (mt/id) :attribute "impersonation_attr_1"}]
+                                                :attributes     {"impersonation_attr_1" "impersonation_role_1"}}
+      (advanced-perms.api.tu/with-impersonations {:impersonations [{:db-id (mt/id) :attribute "impersonation_attr_2"}]
+                                                  :attributes     {"impersonation_attr_2" "impersonation_role_2"}}
+        (is (thrown-with-msg?
+             clojure.lang.ExceptionInfo
+             #"Multiple conflicting connection impersonation policies found for current user"
+             (@#'impersonation/connection-impersonation-role (mt/db)))))))
+
+  (testing "Returns nil for superuser, even if they are in a group with an impersonation policy defined"
+    (advanced-perms.api.tu/with-impersonations {:impersonations [{:db-id (mt/id) :attribute "impersonation_attr"}]
+                                                :attributes     {"impersonation_attr" "impersonation_role"}}
+      (is (= "impersonation_role"
+             (@#'impersonation/connection-impersonation-role (mt/db)))))))
+
+(deftest conn-impersonation-test-postgres
+  (mt/test-driver :postgres
```

Comment:
These driver tests for Postgres/Snowflake are both in a single EE test namespace so that they can use the `with-impersonations` macro. There wasn't a perfect way to organize this but this seemed like the best option. If we add impersonation for a lot more drivers in the future maybe we could break out separate namespaces for each driver.